pennylane_lightning/src/lightning_qubit.hpp

```diff
@@ -348,14 +333,99 @@ VectorXcd apply_ops_2q(
         qubit_positions = calculate_qubit_positions(tensor_indices);
 
         if (w.size() == 1) {
-            tensor_contracted = contract_1q_op<State_2q> (evolved_tensor, op_string, wires_to_contract, p);
+            tensor_contracted = contract_1q_op<State_Xq<2>> (evolved_tensor, op_string, wires_to_contract, p);
         }
         else if (w.size() == 2) {
-            tensor_contracted = contract_2q_op<State_2q> (evolved_tensor, op_string, wires_to_contract, p);
+            tensor_contracted = contract_2q_op<State_Xq<2>> (evolved_tensor, op_string, wires_to_contract, p);
         }
         evolved_tensor = tensor_contracted;
     }
-    State_2q shuffled_evolved_tensor = evolved_tensor.shuffle(qubit_positions);
+    State_Xq<2> shuffled_evolved_tensor = evolved_tensor.shuffle(qubit_positions);
 
     return Map<VectorXcd> (shuffled_evolved_tensor.data(), shuffled_evolved_tensor.size(), 1);
 }
+
+// Template classes for a generic interface for qubit operations
+
+template<int Dim, int ValueIdx>
+class QubitOperationsGenerator
+{
+public:
+    template<typename... Values>
+    static inline VectorXcd apply(
+        Ref<VectorXcd> state,
+        const vector<string>& ops,
+        const vector<vector<int>>& wires,
+        const vector<vector<float>>& params,
+        Values... values)
+    {
+        return QubitOperationsGenerator<Dim, ValueIdx - 1>::apply(state, ops, wires, params, 2, values...);
```

Comment:
The `ValueIdx==0` case is handled by the specialisation:

```
// Generic multi-qubit case
template<int Dim>
class QubitOperationsGenerator<Dim, 0>
```

which is how the recursive template terminates. I could add a runtime check here, but it seems unnecessary especially since it's going to be checked `N` times if `ValueIdx` starts at `N` due to recursion. Unless there's a way of statically enforcing `ValueIdx >= 0`?

Reply:
Right! Was curious about the `ValueIdx==0` and `return QubitOperationsGenerator<Dim, -1>::apply(state, ops, wires, params, 2, values...);` case (and further negative values as the second template argument).

> it seems unnecessary especially since it's going to be checked N times if ValueIdx starts at N due to recursion

Yeah, the current way of using it wouldn't result in `ValueIdx <1`. Was curious what would happen if such a `ValueIdx` is inputted explicitly.

> Unless there's a way of statically enforcing ValueIdx >= 0?

Hmm, think a runtime check would be applicable in such a case. :thinking:

Reply:
> Was curious what would happen if such a ValueIdx is inputted explicitly.

Because templates are all determined at compile-time, I believe it'll recurse up until the compiler's template recursion limit, at which point it'll fail to compile. So in theory, it shouldn't even compile.

Of course, we might at some point hit the template recursion limit for really large numbers of qubits. At least on MSVC, its larger than 50. If we do cross that bridge, a more scalable approach will be needed that doesn't involve Eigen tensors with an arbitrary rank.

Reply:
That's right, just checked it and it indeed runs into compile-time problems. Thanks for the clarification!